### PR TITLE
Dockerfile: Use Multi-Stage Build

### PR DESCRIPTION
**Pull request by mostafahussein**
This PR updates the current Dockerfile to use Multi-Stage build.
@FireFart I tried to avoid same issues that i went through while updating wpscan's dockerfile. let me know if there is something i should modify

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -44,15 +30,26 @@ RUN apk update && \
     && echo "gem: --no-ri --no-rdoc" > /etc/gemrc \
     && gem update --system \
     && gem install bundler \
-    && bundle install --system $BUNDLER_ARGS \
-    && apk del .ruby-builddeps \
-    && rm -rf /var/cache/apk/*
+    && bundle install --system $BUNDLER_ARGS
```
Comment (FireFart) [inline Dockerfile]: can you please add the `--clean` and `--no-cache ` parameters to bundle install? This should clean stuff up afterwards (not really related to this change but some nice addons)

File: Dockerfile [hunk 2]
```diff
@@ -44,15 +30,26 @@ RUN apk update && \
...
 
+COPY . $APP_HOME
```
Comment (FireFart) [inline Dockerfile]: this line should be removed

File: Dockerfile [hunk 3]
```diff
@@ -44,15 +30,26 @@ RUN apk update && \
...
+
+FROM ruby:2.5.1-alpine3.7
+LABEL maintainer="Rapid7"
+
+ENV APP_HOME /usr/src/metasploit-framework/
+ENV NMAP_PRIVILEGED=""
+
+COPY --from=builder /usr/local/bundle /usr/local/bundle
+COPY --from=builder $APP_HOME $APP_HOME
```
Comment (FireFart) [inline Dockerfile]: this should be `COPY . $APP_HOME`

**Review by firefart — CHANGES_REQUESTED** [review]
Thanks for the PR! Looks good so far, only had some small changes

**Comment by mostafahussein** [comment]
@FireFart , I am made the requested modifications

File: Dockerfile [hunk 4]
```diff
@@ -44,15 +30,24 @@ RUN apk update && \
     && echo "gem: --no-ri --no-rdoc" > /etc/gemrc \
     && gem update --system \
     && gem install bundler \
-    && bundle install --system $BUNDLER_ARGS \
-    && apk del .ruby-builddeps \
-    && rm -rf /var/cache/apk/*
+    && bundle install --clean --no-cache --system $BUNDLER_ARGS
```
Comment (FireFart) [inline Dockerfile]: Had a look at the generated image and it looks like there is currently an issue with bundler

Can you please add the following line to this command?

```
  # temp fix for https://github.com/bundler/bundler/issues/6680
  rm -rf /usr/local/bundle/cache
```

**Review by firefart — CHANGES_REQUESTED** [review]

**Review by firefart — APPROVED** [review]

**Comment by FireFart** [comment]
Thanks for the changes!

**Comment by FireFart** [comment]
# Release notes

This change updates the docker image to make use of multi stage builds

**Pull request merged**